plone/app/vocabularies/__init__.py

```diff
@@ -32,3 +34,28 @@ def __getitem__(self, start, stop=None):
 
     def __len__(self):
         return len(self._terms)
+
+
+@implementer(IPermissiveVocabulary)
+class PermissiveVocabulary(SimpleVocabulary):
+    """
+    Permissive vocabulary for cases of integer-keyed choices or cases
+    where vocabulary may mutate later in a transaction to include a
+    newly inserted value.
+    """
+
+    def __contains__(self, value):
+        return True
+
+    def getTermByToken(self, token):
+        """
+        this works around z3c.form.widget.SequenceWidget.extract()
+        pseudo-validation (which is broken for a permissive vocabulary).
+        """
+        try:
+            v = super(PermissiveVocabulary, self).getTermByToken(token)
+        except LookupError:
+            # fallback using dummy term, assumes token==value
+            return SimpleTerm(token)
```

Comment:
In such case, we are never actually sure what the intended value type is -- a term value can be any of int, unicode, or bytes, depending on context (excluding more exotic term.value types not relevant here).  It is safe to do simply keep the (presumed bytes) token as a value.

However, we could normalize a token into a title in this generated (and temporary) term.  We cannot assume that a token is anything more than 7-bit ASCII (per description in zope.schema.interfaces.ITokenizedTerm).  We can also try to normalize sane unicode from ASCII by `urllib.unquote(token).decode('utf8')` (Python 3 equivalent is `urllib.parse(token)`).

My recommendation is that I leave the value as a token, but try to normalize a title:

```python
_token_parse_py3 = getattr(urllib, 'parse', None)
_token_parse_py27 = lambda token: urllib.unquote(token).decode('utf8')
_token_parse = _token_parse_py3 or _token_parse_py2
parse = lambda token: _token_parse(token)

#...
SimpleTerm(token, title=parse(token))
```